### ISSUE TITLE
Inconsistence in metadata manipulation

### ISSUE THREAD

**Issue by Reto911** [datenlord/datenlord#419]
# TL;DR
Metadata operations are not atomic, therefore, inconsistence will happen in high-race scenarios.

# Details

About two months ago, I reported a bug in this Issue:

> The application panicked when doing `fstest`, caused by an assertion failure.
> 
> https://github.com/datenlord/datenlord/blob/cc474f6c09fa1fd72c9a8161caa392049e65eba3/src/async_fuse/memfs/s3_node.rs#L628
> 
> This bug seems to occur on I/O operations on the mount point of `fuse` filesystem randomly, and I think there are no means to reproduce the bug stably for now. As far as I have seen, it occurred twice in almost 60 `fstest` iterations, and occasionally occurred once when I create a file in the mount point manually.
> 
> The error messages are below:
> ```
> thread 'tokio-runtime-worker' panicked at src/async_fuse/memfs/s3_node.rs:628:9:
> assertion failed: self.open_count.load(Ordering::Acquire) > 0
> stack backtrace:
>   ...
>   17:     0x5609db785b0c - <datenlord::async_fuse::memfs::s3_node::S3Node<S> as datenlord::async_fuse::memfs::node::Node>::dec_open_count::hc5ab37cbbe348afc
>             at /home/shiroan/code/datenlord/src/async_fuse/memfs/s3_node.rs:628:9
>   18:     0x5609db7a460e - <datenlord::async_fuse::memfs::s3_node::S3Node<S> as datenlord::async_fuse::memfs::node::Node>::close::{{closure}}::h70ca4a64e309941f
>             at /home/shiroan/code/datenlord/src/async_fuse/memfs/s3_node.rs:1257:9
>   19:     0x5609dc6acaf9 - <core::pin::Pin<P> as core::future::future::Future>::poll::h3359b61c6bc52dd6
>             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/future/future.rs:125:9
>   20:     0x5609db2cda16 - <datenlord::async_fuse::memfs::s3_metadata::S3MetaData<S> as datenlord::async_fuse::memfs::metadata::MetaData>::release::{{closure}}::{{closure}}::hcab36b4decf61511
>             at /home/shiroan/code/datenlord/src/async_fuse/memfs/s3_metadata.rs:94:37
>   21:     0x5609db0b3f2f - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::h554e00565fb14445
>             at /home/shiroan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.40/src/instrument.rs:321:9
>   22:     0x5609db2ca892 - <datenlord::async_fuse::memfs::s3_metadata::S3MetaData<S> as datenlord::async_fuse::memfs::metadata::MetaData>::release::{{closure}}::h4edfcca73a60b793
>             at /home/shiroan/code/datenlord/src/async_fuse/memfs/s3_metadata.rs:86:5
>   23:     0x5609dc6acaf9 - <core::pin::Pin<P> as core::future::future::Future>::poll::h3359b61c6bc52dd6
>             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/future/future.rs:125:9
>   24:     0x5609db1684bd - <datenlord::async_fuse::memfs::MemFs<M> as datenlord::async_fuse::fuse::file_system::FileSystem>::release::{{closure}}::h9e4ff85bb33cf3ea
>             at /home/shiroan/code/datenlord/src/async_fuse/memfs/mod.rs:732:14
>   25:     0x5609db8a356d - <core::pin::Pin<P> as core::future::future::Future>::poll::h0940f0792282ca78
>             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/future/future.rs:125:9
>   26:     0x5609db4031b8 - datenlord::async_fuse::fuse::session::dispatch::{{closure}}::h9f034daf6a5931fc
>             at /home/shiroan/code/datenlord/src/async_fuse/fuse/session.rs:700:18
>   ...
> ```

Now what caused this problem has been found. As far as we know, almost all metadata operations in `S3Metadata` performs as below:

1. Load (deserialize) one or more `S3Node`s from KV.
2. Modify them, such as increasing the `open_count`, removing one of its childs from it, setting its size, etc.
3. Srote (serialize) them to KV.

Noticed that these operations may be concurrent and are not atomic, which caused inconsistence in the metadata when the times of multiple operations on the same file are overlapped.

Let's think about such a scenario: there was an *OPEN* (called $A$) operation before, which set the variable `open_count` to 1; a *RELEASE* operation for $A$ (called $A'$) happens, then an *OPEN* operation (called $B$) also happens; after $B$, a *RELEASE* operation for $B$ (called $B'$) will also happen. Of course these operations are on the same file. There may be 4 situations in this scenario:

1. Sequentialized
   
   ![1](https://github.com/datenlord/datenlord/assets/37030602/6e35b562-0704-4305-a8ec-891b4144dec9)

   The 3 operations happen one by one, which is the ideal thing. In this case, all variables are handled properly.
2. $A'$ finishes between the operation $B$
   
   ![2](https://github.com/datenlord/datenlord/assets/37030602/a19e4bed-33d6-47cb-be72-53971a808955)

   In this case, $B$ reads `open_count` to be 1 as $A'$ haven't set it as 0 yet. $B$ increases it, and set it to KV. $B'$ will get `open_count` as 2 when it begins, decreases it and set it to KV. Now the `open_count` is set to 1 and will never decrease to 0 in normal cases, which will cause resources leakage.
3. $B$ performs during $A'$
   
   ![3](https://github.com/datenlord/datenlord/assets/37030602/62461771-c057-47d2-a568-5b4418db74b5)

   This is what happened exactly in the former bug report. In this case, $B$ reads `open_count` to be 1 as $A'$ haven't set it as 0 yet. $B$ increases it, and set it to KV. $A'$ set `open_count` to 0 and overwrote it in KV, thus $B'$ will get `open_count` as 0, when it trys to decrease it by 1, the assertion failure happens.
4. $B$ and $B'$ performs during the whole $A'$
   
   ![4](https://github.com/datenlord/datenlord/assets/37030602/82be7227-ab2b-4e10-88c3-23777cf1a604)

   Such a strang case, that the `open_count` variable seems to be set to a normal result. This seems to be a kind of "justice in result". However, not only `open_count` makes differences, this causes, that all changes to metadata during $B$ and $B'$ are invisible after $A'$ finishes. 

This is actually a very serious problem, which has great impact on the correctness of our system. We pay attention to it.

**Comment by Reto911** [comment]
Related PRs:
- #444

**Comment by Reto911** [comment]
#402
